designs/Generic_Builder_Interface.md

```diff
@@ -0,0 +1,92 @@
+
+Taking a dependency on Docker is justified as follows:
+* `aws-lambda-builders` already leverages Docker to build native code compatible with Lambda
+* CDK already depends on Docker to build and push images to ECR, so no extra dependencies are required there.
+* Docker successfully encapsulates the complexity of OS platforms and tool installations, allowing us to scale to a wide variety of use-cases without burdening our customers with toolchain dependencies.
+
+Success criteria for the change
+-------------------------------
+A single library and CLI exists, encapsulating the logic to build deployable artifacts from arbitrary code workspaces. It should attempt to infer boiler-plate details such as maven vs gradle for building java, requiring only the language target from the caller. "This is a java 1.8 project, build it for me".
+
+This CLI may be installed directly into the developer's environment or consumable as encapsulated Docker images specific to various langauges/environments.
+
+User Experience Walkthrough
+---------------------------
+
+*TODO: More details/use-cases*
+
+For the CDK use-case, we wish to instantiate a language-specific function type and only have to point to the project's location:
+
+```ts
+const myFunction = new lambda.JavaFunction(stack, 'MyJavaFunction', {
+  projectPath: './lambda'
+});
+```
+
+The CLI command `cdk synth` or `cdk deploy` will run a docker command to build the project into the `cdk.out` folder. In this example, we know to select the image for Java 1.8 because we know the function's runtime. 
+
+The payload is a stripped down version of the existing RPC format,
+```json
+{
+  "jsonrpc": "2.0",
+  "method": "LambdaBuilder.build",
+  "id": 1,
+  "params": {
+    "__protocol_version": "0.1",
+    "capability": {
+      "language": "java1.8", // signal from the caller, known because we know the function's runtime
+      // don't specify the package manager, infer it in the builder tool
+    },
+    "source_dir": "/path/to/source",
+    "artifacts_dir": "/path/to/cdk.out/",
+    "scratch_dir": "/path/to/tmp",
+  }
+}
+```
+
+Implementation
+==============
+
+Design
+------
+
+*TODO: Explain how this feature will be implemented. Highlight the components
+of your implementation, relationships* *between components, constraints,
+etc.*
+
+Open Issues
+-----------
+1. Do we want to rename this project or create a new package encapsulating the generic functionality?
+2. What to do about `pip install`? Is there a better way to expose the CLI for local installation or inclusion in a `Dockerfile`?
+3. Which environments/configurations should we publish Docker containers for? E.g. one container for NPM, one for Java, one for Python, etc. etc.
+4. Do we want to provide hooks for developers to customize a build when their use-case is not supported?
```

Comment:
Lack of customization hooks seems like a nonstarter. What if a project build requires additional installed tool dependencies, or private package repositories, or ANYTHING, really. Let's give people the ability to build what they need from the start, and add incremental sugar later.

Reply:
customization hooks are usually built into the language specific build system. Maven has like a million hooks, npm has hooks for pre-install post-install, etc.

Reply:
@rix0rrr #4 is tracking passing arbitrary options to build tools. We didn't enable this from the beginning because we wanted builds to work out-of-box with majority of cases without additional work.